{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "record = SeqIO.read(\"MPMM01_MPMM_027.gbk\", \"genbank\")\n",
    "\n",
    "gd_diagram = GenomeDiagram.Diagram(record.id)\n",
    "gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "for feature in record.features:\n",
    "    if feature.type != \"gene\":\n",
    "        #Exclude this feature\n",
    "        continue\n",
    "    if len(gd_feature_set) % 2 == 0:\n",
    "        color = colors.blue\n",
    "    else:\n",
    "        color = colors.lightblue\n",
    "    gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                               color=color, label=True,\n",
    "                               label_size = 14, label_angle=0)\n",
    "\n",
    "#I want to include some strandless features, so for an example\n",
    "#will use EcoRI recognition sites etc.\n",
    "for site, name, color in [(\"GAATTC\",\"EcoRI\",colors.green),\n",
    "                          (\"CCCGGG\",\"SmaI\",colors.orange),\n",
    "                          (\"AAGCTT\",\"HindIII\",colors.red),\n",
    "                          (\"GGATCC\",\"BamHI\",colors.purple)]:\n",
    "    index = 0\n",
    "    while True:\n",
    "        index  = record.seq.find(site, start=index)\n",
    "        if index == -1 : break\n",
    "        feature = SeqFeature(FeatureLocation(index, index+len(site)))\n",
    "        gd_feature_set.add_feature(feature, color=color, name=name,\n",
    "                                   label=True, label_size = 10,\n",
    "                                   label_color=color)\n",
    "        index += len(site)\n",
    "\n",
    "gd_diagram.draw(format=\"linear\", pagesize='A4', fragments=4,\n",
    "                start=0, end=len(record))\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_linear_nice.pdf\", \"PDF\")\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_linear_nice.eps\", \"EPS\")\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_linear_nice.svg\", \"SVG\")\n",
    "\n",
    "gd_diagram.draw(format=\"circular\", circular=True, pagesize=(20*cm,20*cm),\n",
    "                start=0, end=len(record), circle_core = 0.5)\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_circular_nice.pdf\", \"PDF\")\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_circular_nice.eps\", \"EPS\")\n",
    "gd_diagram.write(\"/home/minion/projects/genebanks/plasmid_circular_nice.svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc262\n",
    "http://wilkelab.org/classes/SDS348/2015_spring_discussions/discussion_4.1.15.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the required components of BioPython\n",
    "# Entrez allows us to collect data from NCBI\n",
    "# SeqIO allows us to input and output sequence files\n",
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "Entrez.email = \"agb214@exeter.ac.uk\" # define email to keep the NIH happy\n",
    "\n",
    "# NCBI ID for human hemoglobin\n",
    "hemo_id = \"NM_000558.4\"\n",
    "\n",
    "# Fetch a record from NCBI\n",
    "hemo_handle = Entrez.efetch(db=\"nucleotide\", id=hemo_id, rettype=\"gb\", retmode=\"text\")\n",
    "\n",
    "# Parse the handle\n",
    "hemo_record = SeqIO.read(hemo_handle, \"genbank\")\n",
    "\n",
    "# Use a print statement (!!!) to make sure we actually fetched the record properly\n",
    "print(hemo_record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "directory = '/home/minion/projects/genebanks/viral_contigs/gbk_files'\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith(\".gbk\"):\n",
    "        new_name=os.path.splitext(filename)[0]\n",
    "        print(filename)\n",
    "        record = SeqIO.read(filename, \"genbank\") \n",
    "        gd_diagram = GenomeDiagram.Diagram(new_name)\n",
    "        gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "        gd_feature_set = gd_track_for_features.new_set()\n",
    "        \n",
    "        for feature in record.features:\n",
    "            if feature.type != \"gene\":\n",
    "                continue\n",
    "            if len(gd_feature_set) % 2 == 0:\n",
    "                color = colors.blue\n",
    "            else:\n",
    "                color = colors.lightblue\n",
    "            gd_feature_set.add_feature(feature, sigil=\"ARROW\", color=color, label=True, label_size = 10)\n",
    "            gd_diagram.draw(format=\"linear\", orientation=\"landscape\", pagesize='A1',fragments=25, start=0, end=200000)\n",
    "            gd_diagram.write(new_name+\".pdf\", \"PDF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code is part of the Biopython distribution and governed by its\n",
    "# license.  Please see the LICENSE file that should have been included\n",
    "# as part of this package.\n",
    "#\n",
    "\n",
    "\"\"\"GenomeDiagram script to mimic Proux et al 2002 Figure 6.\n",
    "\n",
    "You can use the Entrez module to download the 3 required GenBank files\n",
    "\n",
    "This is an extended version of the example in the Biopython Tutorial\n",
    "which produces a GenomeDiagram figure close to Proux et al 2002 Figure 6.\n",
    "\n",
    "See https://doi.org/10.1128/JB.184.21.6026-6036.2002\n",
    "\"\"\"\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.colors import red, grey, orange, green, brown\n",
    "from reportlab.lib.colors import blue, lightblue, purple\n",
    "\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.Graphics.GenomeDiagram import CrossLink\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "name = \"Proux Fig 6\"\n",
    "\n",
    "# As explained in the Biopython Tutorial, these are three phage genomes. The first\n",
    "# two are self-containged GenBank files, but the third phage is integrated into a\n",
    "# bacterial genome, thus we slice the full record and also take the reverse\n",
    "# complement to match the strand orientation of the other two phage:\n",
    "\n",
    "A_rec = SeqIO.read(\"NC_002703.gbk\", \"gb\")\n",
    "B_rec = SeqIO.read(\"AF323668.gbk\", \"gb\")\n",
    "C_rec = SeqIO.read(\"NC_003212.gbk\",\n",
    "                   \"gb\")[2587879:2625807].reverse_complement(name=True)\n",
    "records = dict((rec.name, rec) for rec in [A_rec, B_rec, C_rec])\n",
    "\n",
    "# Here we hard code the gene colors for simiplicity and to match the target image.\n",
    "# In practice you might have an automatic mapping based on the gene annotation\n",
    "# or some other classification:\n",
    "\n",
    "A_colors = [red] * 5 + [grey] * 7 + [orange] * 2 + [grey] * 2 + [orange] \\\n",
    "    + [grey] * 11 + [green] * 4 + [grey] + [green] * 2 + [grey, green] \\\n",
    "    + [brown] * 5 + [blue] * 4 + [lightblue] * 5 + [grey, lightblue] \\\n",
    "    + [purple] * 2 + [grey]\n",
    "B_colors = [red] * 6 + [grey] * 8 + [orange] * 2 + [grey] + [orange] \\\n",
    "    + [grey] * 21 + [green] * 5 + [grey] + [brown] * 4 + [blue] * 3 \\\n",
    "    + [lightblue] * 3 + [grey] * 5 + [purple] * 2\n",
    "C_colors = [grey] * 33 + [green] * 5 + [brown] * 4 + [blue] * 2 \\\n",
    "    + [grey, blue] + [lightblue] * 2 + [grey] * 8\n",
    "\n",
    "# Here we hard code a list of cross-links with percentage identity scores, based\n",
    "# on a manual inspection of the target image (there could be mistakes here).\n",
    "# In practice you might generate the list of cross-mappings from a BLAST report\n",
    "# or similar computational analysis:\n",
    "\n",
    "# Tuc2009 (NC_002703) vs bIL285 (AF323668)\n",
    "A_vs_B = [\n",
    "    (99, \"Tuc2009_01\", \"int\"),\n",
    "    (33, \"Tuc2009_03\", \"orf4\"),\n",
    "    (94, \"Tuc2009_05\", \"orf6\"),\n",
    "    (100, \"Tuc2009_06\", \"orf7\"),\n",
    "    (97, \"Tuc2009_07\", \"orf8\"),\n",
    "    (98, \"Tuc2009_08\", \"orf9\"),\n",
    "    (98, \"Tuc2009_09\", \"orf10\"),\n",
    "    (100, \"Tuc2009_10\", \"orf12\"),\n",
    "    (100, \"Tuc2009_11\", \"orf13\"),\n",
    "    (94, \"Tuc2009_12\", \"orf14\"),\n",
    "    (87, \"Tuc2009_13\", \"orf15\"),\n",
    "    (94, \"Tuc2009_14\", \"orf16\"),\n",
    "    (94, \"Tuc2009_15\", \"orf17\"),\n",
    "    (88, \"Tuc2009_17\", \"rusA\"),\n",
    "    (91, \"Tuc2009_18\", \"orf20\"),\n",
    "    (93, \"Tuc2009_19\", \"orf22\"),\n",
    "    (71, \"Tuc2009_20\", \"orf23\"),\n",
    "    (51, \"Tuc2009_22\", \"orf27\"),\n",
    "    (97, \"Tuc2009_23\", \"orf28\"),\n",
    "    (88, \"Tuc2009_24\", \"orf29\"),\n",
    "    (26, \"Tuc2009_26\", \"orf38\"),\n",
    "    (19, \"Tuc2009_46\", \"orf52\"),\n",
    "    (77, \"Tuc2009_48\", \"orf54\"),\n",
    "    (91, \"Tuc2009_49\", \"orf55\"),\n",
    "    (95, \"Tuc2009_52\", \"orf60\"),\n",
    "]\n",
    "\n",
    "# bIL285 (AF323668) vs Listeria innocua prophage 5 (in NC_003212)\n",
    "B_vs_C = [\n",
    "    (42, \"orf39\", \"lin2581\"),\n",
    "    (31, \"orf40\", \"lin2580\"),\n",
    "    (49, \"orf41\", \"lin2579\"),  # terL\n",
    "    (54, \"orf42\", \"lin2578\"),  # portal\n",
    "    (55, \"orf43\", \"lin2577\"),  # protease\n",
    "    (33, \"orf44\", \"lin2576\"),  # mhp\n",
    "    (51, \"orf46\", \"lin2575\"),\n",
    "    (33, \"orf47\", \"lin2574\"),\n",
    "    (40, \"orf48\", \"lin2573\"),\n",
    "    (25, \"orf49\", \"lin2572\"),\n",
    "    (50, \"orf50\", \"lin2571\"),\n",
    "    (48, \"orf51\", \"lin2570\"),\n",
    "    (24, \"orf52\", \"lin2568\"),\n",
    "    (30, \"orf53\", \"lin2567\"),\n",
    "    (28, \"orf54\", \"lin2566\"),\n",
    "]\n",
    "\n",
    "\n",
    "def get_feature(features, id, tags=(\"locus_tag\", \"gene\", \"old_locus_tag\")):\n",
    "    \"\"\"Search list of SeqFeature objects for an identifier under the given tags.\"\"\"\n",
    "    for f in features:\n",
    "        for key in tags:\n",
    "            # tag may not be present in this feature\n",
    "            for x in f.qualifiers.get(key, []):\n",
    "                if x == id:\n",
    "                    return f\n",
    "    raise KeyError(id)\n",
    "\n",
    "\n",
    "gd_diagram = GenomeDiagram.Diagram(name)\n",
    "feature_sets = {}\n",
    "max_len = 0\n",
    "for i, record in enumerate([A_rec, B_rec, C_rec]):\n",
    "    max_len = max(max_len, len(record))\n",
    "    # Allocate tracks 5 (top), 3, 1 (bottom) for A, B, C\n",
    "    # (empty tracks 2 and 4 add useful white space to emphasise the cross links\n",
    "    # and also serve to make the tracks vertically more compressed)\n",
    "    gd_track_for_features = gd_diagram.new_track(5 - 2 * i,\n",
    "                                                 name=record.name,\n",
    "                                                 greytrack=True, height=0.5,\n",
    "                                                 start=0, end=len(record))\n",
    "    assert record.name not in feature_sets\n",
    "    feature_sets[record.name] = gd_track_for_features.new_set()\n",
    "\n",
    "# We add dummy features to the tracks for each cross-link BEFORE we add the\n",
    "# arrow features for the genes. This ensures the genes appear on top:\n",
    "for X, Y, X_vs_Y in [(\"NC_002703\", \"AF323668\", A_vs_B),\n",
    "                     (\"AF323668\", \"NC_003212\", B_vs_C)]:\n",
    "    features_X = records[X].features\n",
    "    features_Y = records[Y].features\n",
    "    set_X = feature_sets[X]\n",
    "    set_Y = feature_sets[Y]\n",
    "    for score, x, y in X_vs_Y:\n",
    "        color = colors.linearlyInterpolatedColor(colors.white, colors.firebrick,\n",
    "                                                 0, 100, score)\n",
    "        border = colors.lightgrey\n",
    "        f_x = get_feature(features_X, x)\n",
    "        F_x = set_X.add_feature(SeqFeature(FeatureLocation(f_x.location.start,\n",
    "                                                           f_x.location.end,\n",
    "                                                           strand=0)),\n",
    "                                color=color, border=border)\n",
    "        f_y = get_feature(features_Y, y)\n",
    "        F_y = set_Y.add_feature(SeqFeature(FeatureLocation(f_y.location.start,\n",
    "                                                           f_y.location.end,\n",
    "                                                           strand=0)),\n",
    "                                color=color, border=border)\n",
    "        gd_diagram.cross_track_links.append(CrossLink(F_x, F_y, color, border))\n",
    "\n",
    "\n",
    "for record, gene_colors in zip([A_rec, B_rec, C_rec],\n",
    "                               [A_colors, B_colors, C_colors]):\n",
    "    gd_feature_set = feature_sets[record.name]\n",
    "\n",
    "    i = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type != \"gene\":\n",
    "            # Exclude this feature\n",
    "            continue\n",
    "        try:\n",
    "            g_color = gene_colors[i]\n",
    "        except IndexError:\n",
    "            print(\"Don't have color for %s gene %i\" % (record.name, i))\n",
    "            g_color = grey\n",
    "        gd_feature_set.add_feature(feature, sigil=\"BIGARROW\",\n",
    "                                   color=g_color, label=True,\n",
    "                                   name=str(i + 1),\n",
    "                                   label_position=\"start\",\n",
    "                                   label_size=6, label_angle=0)\n",
    "        i += 1\n",
    "\n",
    "gd_diagram.draw(format=\"linear\", pagesize='A4', fragments=1,\n",
    "                start=0, end=max_len)\n",
    "gd_diagram.write(name + \".pdf\", \"PDF\")\n",
    "gd_diagram.write(name + \".eps\", \"EPS\")\n",
    "gd_diagram.write(name + \".svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio import SeqIO\n",
    "\n",
    "wkdir=\"/home/minion/projects/genebanks/prokka_top10/\"\n",
    "records = list(SeqIO.parse(wkdir+\"AAEN.gbk\", \"genbank\"))\n",
    "#record = SeqIO.read(wkdir+\"AAEN.gbk\", \"genbank\")\n",
    "\n",
    "#Create a track, and a diagram\n",
    "gd_diagram = GenomeDiagram.Diagram(\"AAEN\")\n",
    "gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "#Create the feature set and its feature objects,\n",
    "for record in records:\n",
    "    for feature in record.features:\n",
    "        if feature.type != \"gene\":\n",
    "            #Exclude this feature\n",
    "            continue\n",
    "        if len(gd_feature_set) % 2 == 0:\n",
    "            color = colors.blue\n",
    "        else:\n",
    "            color = colors.lightblue\n",
    "        #Now have to glue the bits together...\n",
    "        gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                   color=color, label=True,\n",
    "                                   label_size = 8, label_angle=0)\n",
    "    \n",
    "gd_diagram.draw(format=\"circular\", circular=True, pagesize=(21.0*cm,29.7*cm),\n",
    "                start=0, end=len(record), circle_core=0.7)\n",
    "\n",
    "gd_diagram.write(wkdir+\"AAEN.svg\", \"SVG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "ename": "NotImplementedError",
     "evalue": "SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNotImplementedError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-84-1b0aa7090ee1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mtest\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mfeature\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfeatures\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mtest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/Bio/SeqRecord.py\u001b[0m in \u001b[0;36m__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    749\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__eq__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 750\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mNotImplementedError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_NO_SEQRECORD_COMPARISON\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    751\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    752\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__ne__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNotImplementedError\u001b[0m: SeqRecord comparison is deliberately not implemented. Explicitly compare the attributes of interest."
     ]
    }
   ],
   "source": [
    "test =[]\n",
    "for feature in record.features in records:\n",
    "    test.append(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "418\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
